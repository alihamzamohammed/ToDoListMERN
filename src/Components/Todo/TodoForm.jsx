import React, { useState, useEffect } from "react";
import { useParams, useHistory } from "react-router-dom";
import { Form, Button } from "react-bootstrap";

const TodoForm = () => {
  const { paramId } = useParams();
  const [id, setId] = useState();
  const history = useHistory();
  const [title, setTitle] = useState("");
  const [content, setContent] = useState("");
  const [selectedCategory, setSelectedCategory] = useState("");
  const [possibleCategories, setPossibleCategories] = useState([]);
  const [response, setResponse] = useState("");
  const [disabled, setDisabled] = useState(false);
  const [dateAdded, setDateAdded] = useState("");

  const createTodo = async () => {
    let send = await fetch(`http://localhost:5050/todo/create`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        title: title,
        content: content,
        category:
          selectedCategory !== "0" ? String(selectedCategory) : undefined,
      }),
    });
    console.log(send);
    setResponse("Todo was created successfully!");
  };

  const updateTodo = async () => {
    await fetch(`http://localhost:5050/todo/update/${id}`, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        title: title,
        content: content,
        category:
          selectedCategory !== "0" ? String(selectedCategory) : undefined,
      }),
    });
    setResponse("Todo was updated successfully!");
  };

  const deleteTodo = async () => {
    await fetch(`http://localhost:5050/todo/delete/${id}`, {
      method: "DELETE",
    });
    setResponse("Todo was deleted successfully!");
    setDisabled(true);
  };

  const create = typeof paramId === "undefined" || paramId === null;

  useEffect(() => {
    const readTodo = async () => {
      try {
        const read = await fetch(`http://localhost:5050/todo/read/${id}`);
        const json = await read.json();
        setTitle(json.title);
        setContent(json.content);
        setDateAdded(json.dateAdded);
        if (json.hasOwnProperty("category")) {
          setSelectedCategory(json.category._id);
        } else {
          setSelectedCategory("0");
        }
      } catch (err) {
        console.error(err);
      }
    };
    const readCategories = async () => {
      try {
        const read = await fetch(`http://localhost:5050/category/read`);
        const json = await read.json();
        setPossibleCategories(json);
      } catch (err) {
        console.error(err);
      }
    };
    setId(paramId);
    readCategories();
    if (!create) {
      readTodo();
    } else {
      setSelectedCategory("0");
    }
  }, [create, id, paramId]);

  return create ? (
    <>
      <h1 className="display-3 title">Create Todo:</h1>
      <Form className="text-input">
        <div className="mb-3">
          <Form.Group>
            <Form.Label htmlFor="title">Title</Form.Label>
            <Form.Control
              type="text"
              placeholder="Enter title"
              name="title"
              value={title}
              onChange={(event) => setTitle(event.target.value)}
            ></Form.Control>
          </Form.Group>
          <Form.Group>
            <Form.Label htmlFor="content">Content</Form.Label>
            <Form.Control
              type="text"
              placeholder="Enter content"
              name="content"
              value={content}
              onChange={(event) => setContent(event.target.value)}
            ></Form.Control>
          </Form.Group>
          <Form.Group>
            <Form.Label htmlFor="categorySelect">Category</Form.Label>
            <Form.Control
              as="select"
              value={selectedCategory}
              onChange={(event) => setSelectedCategory(event.target.value)}
            >
              <option value={0}>Unsorted Todos</option>
              {possibleCategories.map((cat, idx) => (
                <option value={cat._id} key={idx}>
                  {cat.name}
                </option>
              ))}
            </Form.Control>
          </Form.Group>
        </div>
        <div className="mb-3 gx-5">
          <Button
            variant="success"
            onClick={() => createTodo()}
            className="form-button"
          >
            Create
          </Button>
          <Button
            variant="primary"
            onClick={() => setTitle("")}
            className="form-button"
          >
            Reset
          </Button>
          <Button
            variant="warning"
            onClick={() => history.push("/")}
            className="form-button"
          >
            Discard
          </Button>
        </div>
      </Form>
      <div>
        <p className="response">{response}</p>
      </div>
    </>
  ) : (
    <>
      <h1 className="display-3 title">Edit Todo:</h1>
      <Form className="text-input">
        <div className="mb-3">
          <p className="text-muted">ID: {id}</p>
          <p className="text-muted">
            Date Added: {new Date(dateAdded).toUTCString()}
          </p>
        </div>
        <div className="mb-3">
          <Form.Group>
            <Form.Label htmlFor="title">Title</Form.Label>
            <Form.Control
              type="text"
              placeholder="Enter title"
              name="title"
              value={title}
              onChange={(event) => setTitle(event.target.value)}
              disabled={disabled}
            ></Form.Control>
          </Form.Group>
          <Form.Group>
            <Form.Label htmlFor="content">Content</Form.Label>
            <Form.Control
              type="text"
              placeholder="Enter content"
              name="content"
              value={content}
              onChange={(event) => setContent(event.target.value)}
              disabled={disabled}
            ></Form.Control>
          </Form.Group>
          <Form.Group>
            <Form.Label htmlFor="categorySelect">Category</Form.Label>
            <Form.Control
              as="select"
              value={selectedCategory}
              onChange={(event) => setSelectedCategory(event.target.value)}
              disabled={disabled}
            >
              <option value={0}>Unsorted Todos</option>
              {possibleCategories.map((cat, idx) => (
                <option value={cat._id} key={idx}>
                  {cat.name}
                </option>
              ))}
            </Form.Control>
          </Form.Group>
        </div>
        <div className="mb-3 gx-5">
          <Button
            variant="success"
            onClick={() => updateTodo()}
            className="form-button"
            disabled={disabled}
          >
            Update
          </Button>
          <Button
            variant="primary"
            onClick={() => setTitle("")}
            className="form-button"
            disabled={disabled}
          >
            Reset
          </Button>
          <Button
            variant="warning"
            onClick={() => history.push("/")}
            className="form-button"
            disabled={disabled}
          >
            Discard
          </Button>
          <Button
            variant="danger"
            onClick={() => deleteTodo()}
            className="form-button"
            disabled={disabled}
          >
            Delete
          </Button>
        </div>
      </Form>
      <div>
        <p className="response">{response}</p>
      </div>
    </>
  );
};

export default TodoForm;
