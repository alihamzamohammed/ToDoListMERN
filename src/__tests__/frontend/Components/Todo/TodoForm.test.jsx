import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import { enableFetchMocks } from "jest-fetch-mock";
import { MemoryRouter as Router, Route } from "react-router-dom";
import TodoForm from "../../../../Components/Todo/TodoForm";

enableFetchMocks();

describe("Category Form Tests", () => {
  let container;

  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  });

  afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
    container = null;
    fetch.resetMocks();
  });

  it("Renders the Edit Todo Form", async () => {
    const testTodo = {
      title: "Test Title",
      content: "Test Content",
      category: { _id: "123" },
      _id: "123",
      dateAdded: "2021-07-06T14:20:56.175Z",
    };

    const testCats = [
      {
        name: "Test Category",
        _id: "123",
        dateAdded: "2021-07-06T14:20:56.175Z",
      },
    ];

    fetch
      .mockResponseOnce(JSON.stringify(testCats))
      .mockResponseOnce(JSON.stringify(testTodo));

    jest.mock("react-router", () => ({
      ...jest.requireActual("react-router"),
      useParams: () => ({
        paramId: "123",
      }),
    }));

    await act(async () => {
      render(
        <Router initialEntries={["/edit/todo/123"]}>
          <Route path="/edit/todo/:paramId">
            <TodoForm></TodoForm>
          </Route>
        </Router>,
        container
      );
    });
    expect(container).toMatchSnapshot();
  });

  it("Renders the Create Todo Form", async () => {
    const testCats = [
      {
        name: "Test Category",
        _id: "123",
        dateAdded: "2021-07-06T14:20:56.175Z",
      },
    ];

    fetch.mockResponse(JSON.stringify(testCats));

    await act(async () => {
      render(
        <Router>
          <TodoForm></TodoForm>
        </Router>,
        container
      );
    });
    expect(container).toMatchSnapshot();
  });
});
