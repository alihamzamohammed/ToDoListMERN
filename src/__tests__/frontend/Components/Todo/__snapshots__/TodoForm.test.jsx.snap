// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category Form Tests Renders the Create Todo Form 1`] = `
<div>
  <h1
    class="display-3 title"
  >
    Create Todo:
  </h1>
  <form
    class="text-input"
  >
    <div
      class="mb-3"
    >
      <div
        class="form-group"
      >
        <label
          class="form-label"
          for="title"
        >
          Title
        </label>
        <input
          class="form-control"
          name="title"
          placeholder="Enter title"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
          for="content"
        >
          Content
        </label>
        <input
          class="form-control"
          name="content"
          placeholder="Enter content"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
          for="categorySelect"
        >
          Category
        </label>
        <select
          class="form-control"
          name="possibleCategories"
        >
          <option
            value="0"
          >
            Unsorted Todos
          </option>
          <option
            value="123"
          >
            Test Category
          </option>
        </select>
      </div>
    </div>
    <div
      class="mb-3 gx-5"
    >
      <button
        class="form-button btn btn-success"
        type="button"
      >
        Create
      </button>
      <button
        class="form-button btn btn-primary"
        type="button"
      >
        Reset
      </button>
      <button
        class="form-button btn btn-warning"
        type="button"
      >
        Discard
      </button>
    </div>
  </form>
  <div>
    <p
      class="response"
    />
  </div>
</div>
`;

exports[`Category Form Tests Renders the Edit Todo Form 1`] = `
<div>
  <h1
    class="display-3 title"
  >
    Edit Todo:
  </h1>
  <form
    class="text-input"
  >
    <div
      class="mb-3"
    >
      <p
        class="text-muted"
      >
        ID: 
        123
      </p>
      <p
        class="text-muted"
      >
        Date Added: 
        Tue, 06 Jul 2021 14:20:56 GMT
      </p>
    </div>
    <div
      class="mb-3"
    >
      <div
        class="form-group"
      >
        <label
          class="form-label"
          for="title"
        >
          Title
        </label>
        <input
          class="form-control"
          name="title"
          placeholder="Enter title"
          type="text"
          value="Test Title"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
          for="content"
        >
          Content
        </label>
        <input
          class="form-control"
          name="content"
          placeholder="Enter content"
          type="text"
          value="Test Content"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
          for="categorySelect"
        >
          Category
        </label>
        <select
          class="form-control"
          name="possibleCategories"
        >
          <option
            value="0"
          >
            Unsorted Todos
          </option>
          <option
            value="123"
          >
            Test Category
          </option>
        </select>
      </div>
    </div>
    <div
      class="mb-3 gx-5"
    >
      <button
        class="form-button btn btn-success"
        type="button"
      >
        Update
      </button>
      <button
        class="form-button btn btn-primary"
        type="button"
      >
        Reset
      </button>
      <button
        class="form-button btn btn-warning"
        type="button"
      >
        Discard
      </button>
      <button
        class="form-button btn btn-danger"
        type="button"
      >
        Delete
      </button>
    </div>
  </form>
  <div>
    <p
      class="response"
    />
  </div>
</div>
`;
